### PR TITLE
wrap type and projection of operators in a struct

### PR DESCRIPTION
**Pull request by jdroenner**
this can also  hold tile size/resolution information for raster and fields for vector data

### CODE REVIEW

File: operators/src/mock/mock_raster_point_join.rs [hunk 1]
```diff
@@ -127,18 +119,34 @@ impl VectorOperator for MockRasterPointJoinOperator {
...
+        if self.result_descriptor.is_none() {
+            let new_result_descriptor = VectorResultDescriptor {
+                projection: self.point_sources.get(0).map_or_else(
+                    || ProjectionOption::None,
+                    |o| o.result_descriptor().projection,
+                ),
+                data_type: VectorDataType::MultiPoint,
+            };
+
+            // self.result_descriptor = Some(new_result_descriptor); //TODO: discuss if this should be interior or if method is mut...
```
Comment (jdroenner) [inline operators/src/mock/mock_raster_point_join.rs:132]: to enable storage of the computed information, we need to change the result_descriptor method take &mut self. Is there any reason to not do this?

File: operators/src/engine/operator.rs [hunk 2]
```diff
@@ -65,21 +43,236 @@ pub trait VectorOperator: Operator + CloneableVectorOperator {
...
+    /// Wrap a box around a `RasterOperator`
+    fn boxed(self) -> Box<dyn RasterOperator>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+pub type ExecutionContext = u32;
```
Comment (ChristianBeilschmidt) [inline operators/src/engine/operator.rs:60]: Should be an empty struct I guess. Isn't the context something rather for the query method?

File: operators/src/engine/operator.rs [hunk 3]
```diff
@@ -65,21 +43,236 @@ pub trait VectorOperator: Operator + CloneableVectorOperator {
...
+        self.as_ref().vector_processor()
+    }
+}
+
+pub struct InitilaizedOperatorImpl<Parameters, ResultDescriptor> {
+    pub params: Parameters,
+    pub raster_sources: Vec<Box<dyn InitializedRasterOperator>>,
+    pub vector_sources: Vec<Box<dyn InitializedVectorOperator>>,
+    pub context: ExecutionContext,
+    pub result_descriptor: ResultDescriptor,
```
Comment (ChristianBeilschmidt) [inline operators/src/engine/operator.rs:132]: Why is it not necessary to have a common trait for the result description or sufficient to just use one struct?

File: operators/src/engine/operator.rs [hunk 4]
```diff
@@ -65,21 +43,236 @@ pub trait VectorOperator: Operator + CloneableVectorOperator {
...
+    }
+
+    pub fn map_projection<F>(mut self, f: F) -> Self
+    where
+        F: Fn(ProjectionOption) -> ProjectionOption,
+    {
+        self.projection = f(self.projection);
+        self
+    }
+}
```
Comment (ChristianBeilschmidt) [inline operators/src/engine/operator.rs:274]: Maybe we should put this logic into a separate file

File: operators/src/mock/mock_raster_point_join.rs [hunk 5]
```diff
@@ -62,91 +63,101 @@ where
...
+        .map(InitilaizedOperatorImpl::boxed)
     }
+}
 
+impl InitializedVectorOperator
+    for InitilaizedOperatorImpl<MockRasterPointJoinParams, VectorResultDescriptor>
+{
     fn vector_processor(&self) -> Result<crate::engine::TypedVectorQueryProcessor> {
-        self.validate_children(1..2, 1..2)?;
+        // self.validate_children(self.result_descriptor().projection, 1..2, 1..2)?;
```
Comment (ChristianBeilschmidt) [inline operators/src/mock/mock_raster_point_join.rs:116]: Why is this commented out?

File: operators/src/source/gdal_source.rs [hunk 6]
```diff
@@ -391,54 +392,66 @@ where
...
+    fn initialized_operator(
+        self: Box<Self>,
+        context: crate::engine::ExecutionContext,
+    ) -> Result<Box<dyn InitializedRasterOperator>> {
+        InitilaizedOperatorImpl::create(
+            self.params.clone(),
+            context,
+            |params, _, _, _| {
+                let dataset_information =
+                    JsonDatasetInformationProvider::with_dataset_id(&params.dataset_id)?; // TODO: keep the information somewhere
```
Comment (ChristianBeilschmidt) [inline operators/src/source/gdal_source.rs:408]: Create state field in operator?

File: operators/src/engine/operator.rs (hunk 2 above)
Comment (jdroenner) [reply operators/src/engine/operator.rs:60]: this context is required for information like data location and user information. One use-case is the following: Some of the sources need to check with a database and /or config file. Using this context, we can keep the config management separated from the operators.

File: operators/src/engine/operator.rs (hunk 3 above)
Comment (jdroenner) [reply operators/src/engine/operator.rs:132]: The ResultDescriptor is (currently) specific to the kind of operator. A VectorOperator has a VectorResultDescriptor and a RasterOperator a RasterResultDescriptor. We could add a common trait for things like projection.

File: operators/src/engine/operator.rs (hunk 4 above)
Comment (jdroenner) [reply operators/src/engine/operator.rs:274]: ok

File: operators/src/engine/operator.rs (hunk 2 above)
Comment (jdroenner) [reply operators/src/engine/operator.rs:60]: converted it into a struct.

File: operators/src/mock/mock_raster_point_join.rs (hunk 5 above)
Comment (jdroenner) [reply operators/src/mock/mock_raster_point_join.rs:117]: i guess validation should be done in the initialized_operator construction step. This should be deleted.

File: operators/src/source/gdal_source.rs (hunk 6 above)
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: done

File: operators/src/engine/operator.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply operators/src/engine/operator.rs]: I guess that trait makes sense to group common functionality, which is for now the projection.

**Pull request merged**